### ISSUE TITLE
[selectors] consider preserving invalid selectors in :is() and :where()

### ISSUE THREAD

**Issue by emilio** [w3c/csswg-drafts#8356]
In #7972 we resolved to preserve invalid selectors that contain an `&` as-is. It seems it'd be more consistent to do that for _all_ invalid `:is()` / `:where()` branches like `:is(:unknown)`. That's how `@media` and `@supports` behave.

**Comment by Loirooriol** [comment]
If invalid selectors are no longer dropped, `:is()` without arguments does no longer seem needed.

**Comment by tabatkins** [comment]
Yeah I think I'm fine with this. It certainly makes things more consistent, which is nice.

**Comment by cdoublev** [comment]
Preserving invalid/unknown `<media-query>` instead of replacing them with `not all` would also makes things more consistent.

**Comment by cdoublev** [comment]
> If invalid selectors are no longer dropped, `:is()` without arguments does no longer seem needed.

`:is(), valid {}` would become invalid, isn't it? Is it ok regarding back-compatibility?

---

  > TabAtkins: not sure why we decided to drop the invalid selector arguments in `:is()`

I think a good reason is that the author can see which selectors are invalid.

---

  > emilio: implementation-wise, it's easier to just drop it, but if we need to preserve it, easier to preserve always rather than sometimes

Isn't it also easier to serialize all selectors as is? 

You parse `:is(<declaration-value>#)` and then you evaluate each `<declaration-value>` with *parse a comma-separated list according to a CSS grammar* (`<complex-selector>`). This could be applied to `<media-query-list>`.

**Comment by emilio** [comment]
> > If invalid selectors are no longer dropped, `:is()` without arguments does no longer seem needed.
> 
> `:is(), valid {}` would become invalid, isn't it? Is it ok regarding back-compatibility?

Probably, since the only reasonable use-case for `:is()` is "all the other selectors were dropped". But we could keep `:is()` working, if needed for compat.

> Isn't it also easier to serialize all selectors as is?

Not really, because then you need to preserve the specified string. Also it would be a pretty big behavior change (e.g. `.foo   /* whatever */       .bar` would serialize as-is rather than with whitespace normalized and comments removed.

**Comment by cdoublev** [comment]
To be clear, I was suggesting to serialize the specified string of all selectors only in `:is()`/`:where()`. 

On second thought, I also think it is easier to only serialize invalid selectors *as-is*, which I interpret as with the specified string, rather than the component values.

Is it possible to serialize `:is(##invalid, valid)` without a whitespace between `#` and `#invalid`, without the specified string?

**Comment by tabatkins** [comment]
> Is it possible to serialize :is(##invalid, valid) without a whitespace between # and #invalid, without the specified string?

Yeah, assuming you're reserializing from the token stream, that's the behavior you'd get by default. There's no whitespace token there, and the two tokens don't need a separator to re-parse successfully.

**Comment by cdoublev** [comment]
> To parse as a **forgiving selector list** given an input `input`:
  >
  > 1. [Parse a list](https://drafts.csswg.org/css-syntax-3/#css-parse-a-comma-separated-list-according-to-a-css-grammar) of `<complex-real-selector>`s from `input`, and let `selector list` be the result.
  > 2. Remove all failure items from `selector list`, and all items that are invalid selectors, then return a `<selector-list>` representing the remaining items in `selector list`. (This might be empty.)

Does preserving invalid selectors mean that step 2 will be removed and that items in `selector list` represent both the input and the result from its parsing, or that `<forgiving-selector-list>` will no longer be parsed with *parse a comma-separated list according to a CSS grammar*?

**Comment by tabatkins** [comment]
Neither actually - we need to preserve even things that fail to be a `complex-real-selector` at all (currently those will return failure). So I'll have to write something a little differently.

**Comment by sesse** [comment]
A few questions that I hope becomes clear when this actually gets a spec: How do these things re-serialize?

`:is(,,,)` (do we add spaces at opportune points?)
`:host(:is(.a, .b+.c, .d))` (do we keep things that are not parse errors per se, merely disallowed, and do we serialize them back with spaces as if they were fully parsed?)
`:has(:is(:has( .x )))` (very similar issue)

**Comment by byung-woo** [comment]
IMHO, we should preserve the invalid selector text exactly as-is, regardless of why it is invalid (e.g., a pure parsing error, an unallowed complex selector in :host(), or nested :has() restrictions).

We can then apply the existing serialization rule of adding a space after each comma in the list.

Here is how this approach would handle the questions above:
- `:is(,,,)`: The selector list contains four empty strings which are invalid selectors (`:is(<empty>,<empty>,<empty>,<empty>)`). Following the rule, this serializes to `:is(, , , )`.
- `:host(:is(.a, .b+.c, .d))`: The complex selector `.b+.c` is invalid inside `:host()`. By preserving the invalid token as-is, the whole block serializes back to `:host(:is(.a, .b+.c, .d))`.
- `:has(:is(:has(.x)))`: Nested `:has()` selectors are invalid. Preserving the inner invalid selector as-is means this serializes to `:has(:is(:has(.x)))`.

I think this approach provides predictable results for authors and seems to match the Safari and Firefox behavior.

What are your thoughts on this, @sesse?

**Comment by sesse** [comment]
> What are your thoughts on this, [@sesse](https://github.com/sesse)?

I am no longer in Chrome, sorry. (And also on vacation)

**Comment by byung-woo** [comment]
Chrome landed a fix that supports this behind an experimental flag. It also added the following cases to the WPT tests to clarify certain behaviors:

* Preserve empty items, with a whitespace after each comma:
  - `:is(,,,)` -> `:is(, , , )`
* Trim trailing whitespace from an invalid selector:
  - `:is(::before:HOVER  )` -> `:is(::before:HOVER)`
* Serialize an invalid complex/compound selector as-is, even when the selector is valid outside of a forgiving selector list:
  - `:host(:is(.a, .b+.c, .d))`: unchanged after serialization. (No whitespace is added around the `+` combinator, since complex selectors are invalid inside `:host()`).
  - `:is(::before:HOVER, .a)`: unchanged after serialization. (`:HOVER` is not lowercased, since the compound `::before:HOVER` is invalid. `::before` is not allowed inside `:is()`)
  - `:has(:is(:has(.a+.b)))`: unchanged after serialization. (Nested `:has()` is invalid)

**Comment by byung-woo** [comment]
> Neither actually - we need to preserve even things that fail to be a `complex-real-selector` at all (currently those will return failure). So I'll have to write something a little differently.

@tabatkins, Chrome now supports this behavior behind an experimental flag (m153), and the WPT tests now cover the unclear cases raised above, so it seems the spec is ready to be edited. Is there anything else you'd still like to clarify before editing?
